pylint/lint/pylinter.py

```diff
@@ -878,8 +878,12 @@ def _iterate_file_descrs(
 
         The returned generator yield one item for each Python module that should be linted.
         """
+        seen_filepaths: MutableSet[str] = set()  # For deduplication of paths.
         for descr in self._expand_files(files_or_modules):
             name, filepath, is_arg = descr["name"], descr["path"], descr["isarg"]
+            if filepath in seen_filepaths:
```

Comment:
I think we should do that in [``pylint.lint.expand_modules``](https://github.com/PyCQA/pylint/blob/main/pylint/lint/expand_modules.py#L64) directly, maybe change the ``result`` variable to a set, fixing the algorithm directly and possibly de-duplicating ``files_or_modules`` to begin with as soon as we can (directly after parsing, or if it's not reasonable at the beggining of ``expand_modules`` ?). The test could then be added in ``unittest_expand_modules.py``. Doing it after the fact work, but is less efficient :)

Reply:
I'm not sure that we can do this any earlier without the logic becoming messier. If we do this in `expand_modules`, then we have to check and update the seen set in at least two different places (once for packages/subpackages and once for each module file in a package/subpackage). Given that the function is already almost 100 lines long and contains at least 8 branches (not counting loops or try-except suites), I was hesitant to modify it when I inspected it the other day. The seen filepath checks would add another 2 branches. I thought you would probably yell at me if I did it that way. :)

As for returning `result` as a `set`, we would lose order preservation, which could mess with user experience. Instead, we could return it as a `dict`, since those have been order-preserving since Python 3.6. The key could be `filepath` and the value would be the dict that we put into the result list now. I'm not convinced that we would gain efficiency by doing this. We wouldn't save any memory, because the dictionary keys would be equivalent to the seen set. And, we wouldn't save any execution time, because the seen filepath check must look at the same number of filepaths whether it is in `expand_modules` or in `_iter_file_descr`.

Does this analysis sound right to you or did I misunderstand your suggestion?

Reply:
I guess if I implement some variant of your suggestion, `result` will be smaller if duplicates were present. I will make an additional commit, giving this a try.

Reply:
OK, I just pushed an alternative implementation, based on your suggestion.

Reply:
@Pierre-Sassoulas : I think everything is in good shape now. Please let me know what you think.